### PR TITLE
Reduce critical sections in cacher::Watch function

### PR DESCRIPTION
**Pull request by wojtek-t**
Ref https://github.com/kubernetes/kubernetes/issues/73958

### CODE REVIEW

File: staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go [hunk 1]
```diff
@@ -377,20 +363,42 @@ func (c *Cacher) Watch(ctx context.Context, key string, resourceVersion string,
 		chanSize = 1000
 	}
 
+	// Create a watcher here to reduce memory allocations under lock,
+	// given that memory allocation may trigger GC and block the thread.
+	watcher := newCacheWatcher(chanSize, filterWithAttrsFunction(key, pred), emptyFunc, c.versioner)
```
Comment (jpbetz) [inline staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go]: Comments are very helpful in this PR!

Maybe comment here that emptyFunc is a placeholder until the watcher.forget function can be computed?

**Review by jpbetz — APPROVED** [review]
This looks correct to me. Just one minor suggestion.

/lgtm

File: staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go (hunk 1 above)
Comment (wojtek-t) [reply staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go]: SG - https://github.com/kubernetes/kubernetes/pull/75760 out for review (given that this one already merged).

**Pull request merged**